### ISSUE TITLE
Unicode

### ISSUE THREAD

**Issue by ghost** [WebAssembly/WASI#6]
Hi. I would like to propose that all WASI functions require UTF-8 as text encoding and C and C++ bindings be defined using at least `char8_t`.

`char` and `wchar_t` are horrible mistakes that plague not just C but all other languages that take ideas from it. Let's not make this mistake again.

After looking at the spec, I propose this:
`__wasi_string_t` (`struct`)
Members:
* `char8_t *buf` and `size_t buf_len`
The address of the first code unit and amount of code units.

Since `char8_t` has only been accepted to C++20 and not C (but it is expected in C2X), we can create a temporary workaround for C compilers:
`__wasi_char8_t` (`uint8_t`) in C (pre C2X) and `__wasi_char8_t` (`char8_t`) in C++.

Then change the following functions to use `__wasi_string_t`:
* `__wasi_args_get`
* `__wasi_environ_get`
* `__wasi_fd_prestat_dir_name`
* `__wasi_path_create_directory`
* `__wasi_path_filestat_get`
* `__wasi_path_filestat_set_times`
* `__wasi_path_link`
* `__wasi_path_open`
* `__wasi_path_readlink`
* `__wasi_path_remove_directory`
* `__wasi_path_rename`
* `__wasi_path_symlink`
* `__wasi_path_unlink_file`

**Comment by sunfishcode** [comment]
I like this idea!

We will need to continue to support pre-C++2a compilers, but that's doable with the right preprocessor conditionals.

While introducing a struct like `__wasi_string_t` would otherwise be a good idea, in this context, the current wasm ABI for a struct parameter with two members is a pointer argument to a struct in memory, which would make the API considerably uglier at the wasm level.

(This struct ABI will probably change in the future; we may do it at the same time when we add support for returning multi-member structs by value when [multi-value](https://github.com/WebAssembly/multi-value/) becomes available. At that time, I expect we'll also want to convert many of WASI's output parameters into proper return values as well. But until then, it'd be good to avoid having pass-by-value structs in the ABI.)

**Comment by sunfishcode** [comment]
Also, this will depend on the resolution of #8.

**Comment by Lyberta** [comment]
Well, `__wasi_string_t` is similar to `__wasi_iovec_t` so I don't quite see the ugliness being that of a problem.

**Comment by sunfishcode** [comment]
`__wasi_iovec_t` isn't passed or returned by value though; the API uses arrays of `__wasi_iovec_`s, so they're explicitly in memory. It's a subtle distinction, but one practical consideration is that the C struct ABI for by-value arguments and return values may indeed change in the future. While we have a plan to use .o file versioning to help C family users avoid ABI mismatches between .o files compiled from C and family, WASI will want the wasm signature to remain stable, because it'll have users from other languages.

(In the future, I expect we'll move away from using C function signatures as the canonical WASI API description, but that's a separate topic.)

**Comment by pchickey** [comment]
LLVM now has a target for wasm32-wasi, right? What sort of undertaking is it to change the struct parameter ABI for just that target?

**Comment by sunfishcode** [comment]
With the 8.0 release of LLVM, [we documented](http://releases.llvm.org/8.0.0/docs/ReleaseNotes.html#changes-to-the-webassembly-target) the wasm .o file format and C ABI to be stable. It would have been nice to fix this before then, but we ended up prioritizing features needed to allow WASI to launch. We do have a .o versioning mechanism in place specifically to allow us to make changes like this in the future, but it's disruptive, so we're thinking of waiting on struct arguments and changing them at the same time as struct returns, once we have multi-value.

I expect most C users won't be interacting with the WASI APIs directly very often though; they'll typically be using libc, which can take care of all these issues (including casting char* to __wasi_char8_t*).
